mava/distributions.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Optional
+
+import chex
+import jax.numpy as jnp
+import tensorflow_probability.substrates.jax.bijectors as tfb
+import tensorflow_probability.substrates.jax.distributions as tfd
+
+
+class TanhTransformedDistribution(tfd.TransformedDistribution):
+    """A distribution transformed using the `tanh` function."""
+
+    def __init__(
+        self,
+        distribution: tfd.Distribution,
+        bijector: tfb.Bijector = None,
+        threshold: float = 0.999,
+        validate_args: bool = False,
+    ) -> None:
+        """
+        Initialises the TanhTransformedDistribution.
+
+        Args:
+          distribution: The base distribution to be transformed.
+          bijector: The bijective transformation applied to the distribution.
+          threshold: Clipping value for the action when computing the log_prob.
+          validate_args: Whether to validate input with respect to distribution parameters.
+        """
+        if bijector is None:
+            bijector = tfb.Tanh()
+        super().__init__(distribution=distribution, bijector=bijector, validate_args=validate_args)
+        # Computes the log of the average probability distribution outside the
+        # clipping range, i.e. on the interval [-inf, -atanh(threshold)] for
+        # log_prob_left and [atanh(threshold), inf] for log_prob_right.
+        self._threshold = threshold
```

Comment:
Looking at the clipping I *think* that this should be our min/max of actions, but I'm not entirely sure.

Reply:
I think it just keeps you from being allowed to compute arctanh of 1, which will be infinity.